src/ivc/fold_relaxed_plonk_instance_chip.rs

```diff
@@ -0,0 +1,493 @@
+use std::{iter, num::NonZeroUsize};
+
+use ff::{Field, FromUniformBytes, PrimeField, PrimeFieldBits};
+use halo2_proofs::circuit::Value;
+use halo2curves::{Coordinates, CurveAffine};
+use num_traits::Num;
+
+use crate::{
+    constants::NUM_CHALLENGE_BITS,
+    gadgets::{
+        ecc::{AssignedPoint, EccChip},
+        nonnative::bn::{
+            big_uint::{self, BigUint},
+            big_uint_mul_mod_chip::{self, BigUintMulModChip, ModOperationResult},
+        },
+    },
+    main_gate::{AssignedBit, AssignedValue, MainGate, MainGateConfig, RegionCtx, WrapValue},
+    nifs::CrossTermCommits,
+    plonk::{PlonkInstance, RelaxedPlonkInstance},
+    poseidon::ROCircuitTrait,
+    util,
+};
+
+pub(crate) struct FoldRelaxedPlonkInstanceChip<C: CurveAffine>
+where
+    C::Base: PrimeFieldBits + FromUniformBytes<64>,
+{
+    W: C,
+    E: C,
+    u: C::ScalarExt,
+    X0: BigUint<C::ScalarExt>,
+    X1: BigUint<C::ScalarExt>,
+
+    limb_width: NonZeroUsize,
+    limbs_count: NonZeroUsize,
+}
+
+struct AssignedWitness<C: CurveAffine> {
+    public_params_commit: AssignedValue<C::Base>,
+    W: AssignedPoint<C>,
+    E: AssignedPoint<C>,
+    u: AssignedValue<C::Base>,
```

Comment:
as above, we can change it to challenges:
`challenges: [AssignedValue<C::Base>; num_challenges]`

One thing to keep in mind is that the PlonkInstance only contains `num_challenges - 1` of vars (no `u`). RelaxedPlonkInstance contains `num_challenges` of vars (including `u`).

Reply:
In fact, I don't fully understand why we are mixing `u` and `challanges`. Can you give some context?